litr/src/main/java/com/linkedin/android/litr/io/Camera2MediaSource.kt

```diff
@@ -0,0 +1,253 @@
+
+    /**
+     * Opens the configured CameraDevice so that it's ready for recording. Once available, the
+     * configured Callback will be notified via onDeviceReady. Once this has done, the consumer is
+     * able to start the Transcode session.
+     */
+    @SuppressLint("MissingPermission")
+    @Synchronized
+    fun openCamera() {
+        startBackgroundThread()
+        cameraManager.openCamera(cameraId, cameraStateCallback, backgroundHandler)
+    }
+
+    /**
+     * Stops any in-progress recording, allowing the MediaSource to notify the end of the video
+     * stream. It will also allow any internal resources to be cleaned up.
+     */
+    @Synchronized
+    fun stopRecording() {
+        stopExternal()
+
+        captureSession?.apply {
+            stopRepeating()
+            close()
+        }
+    }
+
+    /**
+     * Once a device has been opened, we can query it's Characteristics to know how to configure
+     * the session.
+     */
+    private fun initDevice(camera: CameraDevice) {
+        cameraDevice = camera
+
+        val characteristics = cameraManager.getCameraCharacteristics(camera.id)
+        val sensorRect = characteristics.get(CameraCharacteristics.SENSOR_INFO_ACTIVE_ARRAY_SIZE)
+        val sensorOrientation = characteristics.get(CameraCharacteristics.SENSOR_ORIENTATION)
+
+        val recordSize = getRecordSize(sensorRect, sensorOrientation)
+        width = recordSize.width
+        height = recordSize.height
+
+        callback?.onDeviceReady(characteristics)
+    }
+
+    /**
+     * Starts the capture session for the given CameraDevice. We will configure the input Surface
+     * provided by the ExternalMediaSource, as well as the optional Preview SurfaceHolder.
+     */
+    private fun startCaptureSession(camera: CameraDevice) {
+        val surfaces = mutableListOf<Surface>()
+        captureRequest = camera.createCaptureRequest(CameraDevice.TEMPLATE_RECORD).apply {
+            // Add the (optional) preview SurfaceHolder.
+            previewSurfaceHolder?.let {
+                addTarget(it.surface)
+                surfaces.add(it.surface)
+            }
+
+            // Add the ExternalMediaSource's input Surface.
+            inputSurface?.let {
+                addTarget(it)
+                surfaces.add(it)
+            }
+
+            // Set a fixed target frame rate. This will match what we'll assume we are encoding too.
+            set(CaptureRequest.CONTROL_AE_TARGET_FPS_RANGE, Range(frameRate, frameRate))
+        }.build()
+
+        camera.createCaptureSession(surfaces, captureStateCallback, backgroundHandler)
+    }
+
+    /**
+     * Since we don't support cropping of the sensor or input surface, we will record to a Surface
+     * that has the same aspect ratio as the camera's sensor.
+     */
+    private fun getRecordSize(sensorRect: Rect?, sensorOrientation: Int?): Size {
+        val aspectRatio = sensorRect?.let { it.height().toFloat() / it.width().toFloat() } ?: (3f / 4f)
```

Comment:
note to self: we should add an enum (or a collection of constants) for typical aspect ratios. Don't have to do this in this PR.